test/parallel_api/algorithm/alg.sorting/sort_by_key_common.h

```diff
@@ -0,0 +1,297 @@
+    using ValT = typename std::iterator_traits<ValIt>::value_type;
+    if constexpr (std::is_same_v<ValT, Particle>)
+    {
+        static_assert(std::is_same_v<KeyT, Particle::energy_type>);
+        std::default_random_engine gen{seed};
+        std::uniform_real_distribution<float> mass_dist(0.0, 1000.0);
+        std::uniform_real_distribution<float> velocity_dist(0.0, 1.0);
+        std::uniform_real_distribution<float> coord_dist(0.0, 1.0);
+        std::generate(vals_begin, vals_begin + vals_n, [&gen, &mass_dist, &velocity_dist, &coord_dist]() {
+            return Particle{mass_dist(gen), velocity_dist(gen), {coord_dist(gen), coord_dist(gen), coord_dist(gen)}};
+        });
+        std::transform(vals_begin, vals_begin + keys_n, keys_begin, [](const auto& particle) { return particle.energy(); });
+    }
+    else
+    {
+        TestUtils::generate_arithmetic_data(keys_begin, keys_n, seed);
+        TestUtils::generate_arithmetic_data(vals_begin, vals_n, seed + 1);
+        // avoid having value duplicates at the same locations as key duplicates for a more robust stability check
+        std::shuffle(vals_begin, vals_begin + keys_n, std::default_random_engine(seed));
+    }
+}
+
+template<typename Policy, typename KeyIt, typename ValIt, typename Size, typename Compare>
+void
+call_sort(Policy&& policy, KeyIt keys_begin, ValIt vals_begin, Size n, Compare compare, StableSortTag)
+{
+    if constexpr (std::is_same_v<Compare, NoComparatorTag>)
+        oneapi::dpl::stable_sort_by_key(policy, keys_begin, keys_begin + n, vals_begin);
+    else
+        oneapi::dpl::stable_sort_by_key(policy, keys_begin, keys_begin + n, vals_begin, compare);
+}
+
+template<typename Policy, typename KeyIt, typename ValIt, typename Size, typename Compare>
+void
+call_sort(Policy&& policy, KeyIt keys_begin, ValIt vals_begin, Size n, Compare compare, UnstableSortTag)
+{
+    if constexpr (std::is_same_v<Compare, NoComparatorTag>)
+        oneapi::dpl::sort_by_key(policy, keys_begin, keys_begin + n, vals_begin);
+    else
+        oneapi::dpl::sort_by_key(policy, keys_begin, keys_begin + n, vals_begin, compare);
+}
+
+template<typename KeyIt, typename ValIt, typename Size, typename Compare>
+void
+call_reference_sort(KeyIt ref_keys_begin, ValIt ref_vals_begin, Size n, Compare)
+{
+    static_assert(std::is_same_v<Compare, NoComparatorTag> || std::is_same_v<Compare, GreaterComparatorTag>);
+    auto first = oneapi::dpl::make_zip_iterator(ref_keys_begin, ref_vals_begin);
+    if constexpr (std::is_same_v<Compare, NoComparatorTag>)
+    {
+        std::stable_sort(first, first + n, [](const auto& lhs, const auto& rhs) {
+            return std::get<0>(lhs) < std::get<0>(rhs);
+        });
+    }
+    if constexpr (std::is_same_v<Compare, GreaterComparatorTag>)
+    {
+        std::stable_sort(first, first + n, [](const auto& lhs, const auto& rhs) {
+            return std::get<0>(lhs) > std::get<0>(rhs);
```

Comment:
it is not consistent (and might be non correct) with `call_sort` logic:
In case of not "NoComparatorTag" `call_sort` sorts data via `compare` passed by a caller,
but `call_reference_sort` doesn't work at all (`static_assert` is triggered).

Reply:
Let me clarify what happens in the code:

If `NoComparator` is passed, `call_sort` calls `dpl::sort_by_key` without comparator, and `call_reference_sort` calls `std::sort` with zipped pairs and std::less applied only to keys, which is an equivalent.

I suppose it is what is expected, unless I misunderstood you.

Reply:
Another thing which confuses: 
line 145: `if constexpr (std::is_same_v<Compare, GreaterComparator>)`
But on line 148, we dont call `GreaterComparator::operator()`....
It looks we dont call GreaterComparator::operator() everywhere. So, why we have structute GreaterComparator with operator() ?

Reply:
`GreaterComparator` is called by `call_sort`. `call_reference_sort` used a custom substitute with the same effect.
I've replaced it with the passed comparator using your previous suggestion.